File: examples/rockset/java/Basic.java
/*
Example program using the official Rockset Java Client with a CrateDB backend.
Here: "Documents" and "Queries" APIs, basic usage.

Usage
=====
- Install dependencies: Run `mvn install`.
- Adjust configuration: Define `ROCKSET_APISERVER` environment variable.
- Run program: Run `mvn compile exec:java`.

Documentation
=============
- https://github.com/rockset/rockset-java-client/blob/master/src/main/java/com/rockset/examples/AddDocumentExample.java
- https://github.com/rockset/rockset-java-client/blob/master/src/main/java/com/rockset/examples/QueryExample.java
- https://docs.rockset.com/documentation/reference/adddocuments
- https://docs.rockset.com/documentation/reference/query
*/
import com.rockset.client.RocksetClient;
import com.rockset.client.model.AddDocumentsRequest;
import com.rockset.client.model.AddDocumentsResponse;
import com.rockset.client.model.CreateCollectionRequest;
import com.rockset.client.model.CreateCollectionResponse;
import com.rockset.client.model.QueryRequest;
import com.rockset.client.model.QueryRequestSql;
import com.rockset.client.model.QueryResponse;
import java.time.Instant;
import java.time.LocalDate;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.Map;

public class Basic {
  public static void main(String[] args) throws Exception {

    // Connect to API endpoint.
    String rockset_apiserver = System.getenv("ROCKSET_APISERVER");
    if (rockset_apiserver == null) { rockset_apiserver = "http://localhost:4243"; }
    String rockset_apikey = System.getenv("ROCKSET_APIKEY");
    if (rockset_apikey == null) { rockset_apikey = "abc123"; }

    RocksetClient rs = new RocksetClient(rockset_apikey, rockset_apiserver);

    // TODO: Implement Collections API mock.
    /*
    CreateCollectionRequest request =
        new CreateCollectionRequest().name("my-first-add-document-collection");
    try {
      CreateCollectionResponse response = rs.collections.create("commons", request);
      System.out.println(response);
      String collectionName = response.getData().getName();
      System.out.println("collectionName:", collectionName);
    } catch (Exception e) {
      e.printStackTrace();
    }
    */

    // Add Documents.
    LinkedList<Object> list = new LinkedList<>();
    Map<String, Object> json = new LinkedHashMap<>();
    json.put("name", "foo");
    json.put("address", "bar");
    json.put("dob", LocalDate.of(2000, 7, 4));
    json.put("event_time", Instant.ofEpochSecond(1000));
    list.add(json);

    AddDocumentsRequest documentsRequest = new AddDocumentsRequest().data(list);
    try {
      AddDocumentsResponse documentsResponse =
          rs.documents.add("commons", "foobar", documentsRequest);
      System.out.println(documentsResponse);
    } catch (Exception e) {
      e.printStackTrace();
    }

    // Submit SQL query.
    QueryRequest request =
        new QueryRequest().sql(new QueryRequestSql().query("SELECT * FROM \"commons\".\"foobar\""));

    try {
      QueryResponse response = rs.queries.query(request);
      System.out.println(response);
    } catch (Exception e) {
      e.printStackTrace();
    }

  }
}

File: examples/rockset/python/basic.py
"""
Example program using the official Rockset Python Client with a CrateDB backend.
Here: "Documents" and "Queries" APIs, basic usage.

Usage
=====
- Install dependencies: Run `pip install 'rockset<2.2'`.
- Adjust configuration: Define `ROCKSET_APISERVER` environment variable.
- Run program: Run `python basic.py`

Documentation
=============
- https://github.com/rockset/rockset-python-client/blob/main/docs/DocumentsApi.md#add_documents
- https://github.com/rockset/rockset-python-client/blob/main/docs/QueriesApi.md#query
- https://docs.rockset.com/documentation/reference/adddocuments
- https://docs.rockset.com/documentation/reference/query
"""

import os
from enum import Enum
from pprint import pprint

from rockset import Configuration, RocksetClient
from rockset.model.query_parameter import QueryParameter
from rockset.model.query_request_sql import QueryRequestSql


# Define endpoint of API Emulator.
class Connections(str, Enum):
    rockset_apiserver = os.environ.get("ROCKSET_APISERVER", "http://localhost:4243")
    rockset_apikey = os.environ.get("ROCKSET_APIKEY", "abc123")


def main():
    # Create an instance of the Rockset client.
    rs_config = Configuration(
        host=Connections.rockset_apiserver,
        api_key=Connections.rockset_apikey,
    )
    rs = RocksetClient(config=rs_config)

    # TODO: Create Collection.
    # rs.Collections.create_kafka_collection()  # noqa: ERA001

    # Add Documents.
    api_response = rs.Documents.add_documents(
        collection="collection_example",
        data=[{"id": "foo", "field": "value"}],
    )
    pprint(api_response)  # noqa: T203

    # Submit SQL query.
    api_response = rs.Queries.query(
        sql=QueryRequestSql(
            parameters=[
                QueryParameter(
                    name="_id",
                    type="string",
                    value="foo",
                ),
            ],
            query="SELECT * FROM commons.collection_example WHERE id = :_id",
        ),
    )
    pprint(api_response)  # noqa: T203


if __name__ == "__main__":
    main()
